### PR TITLE
fix: default values of aggregates in STS

### PR DESCRIPTION
**Pull request by mrtraser**
Add default values of aggregation and granularity for synthetic timeseries if they are not provided.

### CODE REVIEW

File: src/parser/ts/index.ts [hunk 1]
```diff
@@ -15,9 +21,19 @@ const compiledGrammar = Grammar.fromCompiled(grammar);
 export const formQueriesForExpression = async (
   expression: string,
   target: QueryTarget,
-  connector: Connector
+  connector: Connector,
+  options: QueryOptions
```
Comment (polomani) [inline src/parser/ts/index.ts:25]: maybe pass just interval?

File: src/parser/ts/index.ts [hunk 2]
```diff
@@ -48,6 +64,32 @@ const NoTimeseriesFound = (filter: StringMap, expr: string) => {
   );
 };
 
+/**
+ * Injects default values of aggregate and granularity (if it is not provided).
+ * Mutates <query> properties array with default aggregate or granularity values.
```
Comment (polomani) [inline src/parser/ts/index.ts:69]: it doesn't mutate, it returns a copy

Comment (polomani) [reply src/parser/ts/index.ts:69]: well, internally it mutates, but it's not visible outside anyway

File: src/parser/ts/index.ts [hunk 3]
```diff
@@ -48,6 +64,32 @@ const NoTimeseriesFound = (filter: StringMap, expr: string) => {
...
+  parsedData: STSQuery,
+  defaultValues: { aggregate: string; granularity: string }
+) => {
+  const result = cloneDeep(parsedData);
+
+  walk(result, obj => {
+    if (isSTSReference(obj)) {
+      Object.keys(defaultValues).forEach(key => {
+        if (findIndex(obj.query, ['path', key]) === -1) {
+          obj.query.push(STSFilter(key, defaultValues[key]));
```
Comment (polomani) [inline src/parser/ts/index.ts:84]: ;)

File: src/parser/ts/parser.unit.test.ts [hunk 4]
```diff
@@ -449,3 +450,54 @@ describe('check if expression has aggregates', () => {
...
+    const withAggregatesAndGranularity = injectAggregatesToParsed(
+      stsWithAggregateAndGranularity,
+      defaults
+    );
+
+    expect(withAggregatesAndGranularity).toEqual(stsWithAggregateAndGranularity);
+  });
+  it('should not add default values in case of default aggregate is none', () => {
+    const sts = STS([Filter('name', 'name')]);
+    const stsWithInjected = injectAggregatesToParsed(sts, { aggregate: 'none', granularity: '2h' });
```
Comment (polomani) [inline src/parser/ts/parser.unit.test.ts:499]: this should be removed now

File: src/partials/query.editor.html [hunk 5]
```diff
@@ -142,11 +142,24 @@
         </span>
       </info-popover>
     </div>
+    <div class="gf-form">
+      <label class="gf-form-label query-keyword fix-query-keyword">Aggregation</label>
+      <select class="gf-form-input" ng-model="ctrl.target.aggregation" ng-init="ctrl.getInitAggregate()" ng-change="ctrl.refreshData()">
+        <option ng-repeat="fn in ctrl.customTabAggregation" value="{{fn.value}}">{{fn.name}}</option>
+      </select>
+    </div>
+    <div class="gf-form" ng-if="ctrl.target.aggregation && ctrl.target.aggregation !== 'none'">
```
[the remark this answers is not in the document]
Comment (mrtraser) [reply src/partials/query.editor.html:151]: yes, and here is used to show/hide when aggregation is selected or not

File: src/parser/ts/index.ts (hunk 1 above)
Comment (mrtraser) [reply src/parser/ts/index.ts:25]: yep, will change

File: src/parser/ts/index.ts (hunk 2 above)
Comment (mrtraser) [reply src/parser/ts/index.ts:69]: I've mentioned it cause implementation is a bit step away of a functional approach

File: src/parser/ts/parser.unit.test.ts (hunk 4 above)
Comment (mrtraser) [reply src/parser/ts/parser.unit.test.ts:499]: why? to set aggregation to `none` is the only way to fetch real data instead of aggregated

Comment (polomani) [reply src/parser/ts/parser.unit.test.ts:499]: I though we could operate with undefined/null value instead.
but ok, this approach works as well

File: src/cdfDatasource.ts [hunk 6]
```diff
@@ -52,7 +51,7 @@ export function formQueryForItems(
       granularity: granularity || ms2String(options.intervalMs),
     };
   }
-  const limit = calculateDPLimitPerQuery(isAggregated, items.length);
+  const limit = calculateDPLimitPerQuery(items.length);
```
Comment (polomani) [inline src/cdfDatasource.ts:54]: for non-aggregated values in normal timeseries limit is still 100k
so no need to change here

File: src/parser/ts/index.ts [hunk 7]
```diff
@@ -15,9 +21,19 @@ const compiledGrammar = Grammar.fromCompiled(grammar);
...
   expression: string,
   target: QueryTarget,
-  connector: Connector
+  connector: Connector,
+  defaultInterval: string
 ): Promise<DataQueryRequestItem[]> => {
-  const parsed = parse(expression);
+  const rawParsed = parse(expression);
+  const defaultGranularity = target.granularity || defaultInterval;
+  const defaultAggregation = target.aggregation;
```
Comment (polomani) [inline src/parser/ts/index.ts:29]: maybe we just pass all this stuff (`target. granularity, granularity.aggregatetion, interval`) to the `injectAggregatesToParsed ` and hide logic there?
same with `defaultAggregation === 'none'`

Comment (polomani) [reply src/parser/ts/index.ts:29]: end call it `enrichWithDefaultAggregates` or something

Comment (polomani) [reply src/parser/ts/index.ts:29]: so we don't bloat the main function

Comment (mrtraser) [reply src/parser/ts/index.ts:29]: hmm, it makes sense, I'll redo this as well

File: src/parser/ts/parser.unit.test.ts (hunk 4 above)
Comment (mrtraser) [reply src/parser/ts/parser.unit.test.ts:499]: I did this in this way cause we have `none` value in 1st tab, just to keep consistency when we're switching tabs

File: src/cdfDatasource.ts [hunk 8]
```diff
@@ -63,8 +62,8 @@ export function formQueryForItems(
   };
 }
 
-function calculateDPLimitPerQuery(hasAggregates: boolean, queriesNumber: number) {
```
Comment (polomani) [inline src/cdfDatasource.ts:66]: for regular ts limit is still different for aggregated/not aggregated dps

File: src/parser/ts/index.ts (hunk 2 above)
Comment (polomani) [reply src/parser/ts/index.ts:60]: I believe it just gives wrong assumptions to the user, as what matters is inputs/outputs, not internal logic.

File: src/parser/ts/parser.unit.test.ts [hunk 9]
```diff
@@ -492,3 +493,76 @@ describe('parse unary "-" operator', () => {
     })
   );
 });
+describe('check if default aggregation and granularity will be added', () => {
+  const defaults = { aggregate: 'average', granularity: '1h' };
+  const defaultOnlyAggregate = { aggregate: 'average', granularity: '' };
+  const defaultInterval = '6h';
+  it('should add default aggregate and granularity', () => {
+    const stsWithoutAggregation = STS([Filter('name', 'name')]);
+    const withoutAggregates = enrichWithDefaultAggregates(
```
Comment (polomani) [inline src/parser/ts/parser.unit.test.ts:502]:
```suggestion
    const withAggregates = enrichWithDefaultAggregates(
```
?

Comment (mrtraser) [reply src/parser/ts/parser.unit.test.ts]: fixed

File: src/cdfDatasource.ts (hunk 8 above)
Comment (mrtraser) [reply src/cdfDatasource.ts:66]: ah, right, nice catch!

**Pull request merged**